debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/transforms/StringToJSON.java

```diff
@@ -0,0 +1,99 @@
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.Map.Entry;
+
+/**
+ * Debezium Mongo Connector generates the CDC records in String format. Sink connectors usually are not able to parse
+ * the string and insert the document as it is represented in the Source. so a user use this SMT to parse the String 
+ * and insert the MongoDB document in the JSON format..
+ * @param <R> the subtype of {@link ConnectRecord} on which this transformation will operate
+ * @author Sairam Polavarapu
+ */
+public class StringToJSON<R extends ConnectRecord<R>> implements Transformation<R> {
+
+    final ExtractField<R> afterExtractor = new ExtractField.Value<R>();
+    final ExtractField<R> patchExtractor = new ExtractField.Value<R>();
+
+    @Override
+    public R apply(R r) {
+        SchemaBuilder schemabuilder = SchemaBuilder.struct();
+        BsonDocument value = null;
+
+        final R afterRecord = afterExtractor.apply(r);
+
+        if (afterRecord.value() == null) {
+            final R patchRecord = patchExtractor.apply(r);
+            value = BsonDocument.parse(patchRecord.value().toString());
+        } else {
+            value = BsonDocument.parse(afterRecord.value().toString());
+        }
+
+        Set<Entry<String, BsonValue>> keyValues = value.entrySet();
+
+        for (Entry<String, BsonValue> keyValuesforSchema : keyValues) {
+            if(keyValuesforSchema.getKey().toString().equalsIgnoreCase("$set")) {
+                BsonDocument val1 = BsonDocument.parse(keyValuesforSchema.getValue().toString());
+                Set<Entry<String, BsonValue>> keyValuesforSetSchema = val1.entrySet();
+                for (Entry<String, BsonValue> keyValuesforSetSchemaEntry : keyValuesforSetSchema) {
+                    MongoDataConverter.addFieldSchema(keyValuesforSetSchemaEntry, schemabuilder);
+                    }
+                } else {
+                    MongoDataConverter.addFieldSchema(keyValuesforSchema, schemabuilder);
+                    }
+        }
+
+        Schema finalSchema = schemabuilder.build();
+        Struct finalStruct = new Struct(finalSchema);
+
+        for (Entry<String, BsonValue> keyvalueforStruct : keyValues) {
+            if(keyvalueforStruct.getKey().toString().equalsIgnoreCase("$set")) {
+                BsonDocument val1 = BsonDocument.parse(keyvalueforStruct.getValue().toString());
+                Set<Entry<String, BsonValue>> keyvalueforSetStruct = val1.entrySet();
+                for (Entry<String, BsonValue> keyvalueforSetStructEntry : keyvalueforSetStruct) {
+                    MongoDataConverter.convertRecord(keyvalueforSetStructEntry, finalSchema, finalStruct);
+                    }
+                } else {
+            MongoDataConverter.convertRecord(keyvalueforStruct, finalSchema, finalStruct);
+            }
+        }
+
+        return r.newRecord(r.topic(), r.kafkaPartition(), r.keySchema(), r.key(), finalSchema, finalStruct,
```

Comment:
So this propagates the key as is (i.e. String-ified), but the same kind of conversion should be applied as for the value columns. Otherwise, the PK column will always of type TEXT (as its Stringified) instead of the actual column type.

Reply:
I see that the key schema is always being populated as the String even if it is a integer or any other datatype.

Reply:
Yes, it's always a String (as is `after` or `patch` within the event value), but the value encoded within that String is encoded using MongoDB's extended JSON serialization. So if e.g. the key is an int, you'll see `{ "id" : "1234" }`, but if it's String, then you'll see `{ "id" : "\"1234\"" }